### PR TITLE
Issue330 force password generation

### LINKED ISSUE #273: use zxcvbn on the server to optionally enforce secure passwords
**Is your feature request related to a problem? Please describe.**
I'd like to be able to enforce secure passwords without adding security limiting password restrictions like "you must include a number"

**Describe the solution you'd like**
Add a config option for the minimum zxcvbn "number of guesses" allowed. All passwords must exceed this value

**Describe alternatives you've considered**
Default the password to "password" and trust people not to use each others accounts.

**Additional context**
No

### LINKED ISSUE #330: Force server side random generation of new passwords
**Is your feature request related to a problem? Please describe.**
Users often select insecure passwords, either passwords that are dictionary words or duplicated on other sites. Even if a password is "zxcvbn secure" it's difficult to determine if the password has been re-used or will be written down.

If the server generates the password we can at least guarantee it has not been used elsewhere before. If the password is of a very large length 100-300 or so characters we can ensure it's unlikely to be written down and will encourage if not require use of a password manager.

**Describe the solution you'd like**
Have the server generate the password and display it once for copying by the user. Add a configuration option to prevent the user from customizing the password.

**Describe alternatives you've considered**
Use zxcvvbn to enforce passwords that appear random and generated

**Additional context**
No additional context

### PR DESCRIPTION
**Pull request by Serk352**
I added new settings, the first, called usePasswordGeneration, if it's set to true, the users can't longer customize the password, and the BE will create a random password.

Then its the PasswordLenght setting, which is used to set the length of the generated password.

The generated password will be shown to the user in the form, as a noneditable field, this field, has two buttons, one for activate/deactivate the visibility and the other is to add the password to the clipboard.

Also, I added the settings called CheckPasswordScore and MinimumScore.

If CheckPasswordScore  is set to true, the submitted password by the user will be evaluated using the zxcvbn nugget, before it is set as the user's new password.

MinimumScore is to set the minimum score the new password has to achieve at being evaluated (by zxcvbn ) before be set as the new password.

Preview:
PasswordGenerator

Closes #330 .
Closes #273 .

### CODE REVIEW

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs [hunk 1]
```diff
@@ -43,6 +52,42 @@ public class PasswordController : Controller
...
+                    firstWord = char.ToUpper(firstWord[0], CultureInfo.InvariantCulture) + firstWord.Substring(1);
+
+                var secondWord = _words[secondIndex];
+                if (int.Parse(secondIndex.ToString().Substring(0, 1)) < 4)
+                    secondWord = char.ToUpper(secondWord[0], CultureInfo.InvariantCulture) + secondWord.Substring(1);
+
+                password = $"{firstWord}{secondIndex}_{secondWord}{firstIndex}";
+                score = Zxcvbn.MatchPassword(password).Score;
+            }
+            while (score < 4);
```
Comment (geoperez) [inline src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: Add a new setting with the minimum score required.

File: src/Unosquare.PassCore.Web/Startup.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@ namespace Unosquare.PassCore.Web
     using Models;
 #if DEBUG
     using Helpers;
+    using System.Collections.Generic;
```
Comment (geoperez) [inline src/Unosquare.PassCore.Web/Startup.cs]: Remove this line

**Comment by graingert** [comment]
will there be an option for a non passphrase password? We want to force the use of a password manager and make it difficult or unpleasant to write out

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs [hunk 3]
```diff
@@ -43,6 +52,42 @@ public class PasswordController : Controller
         [HttpGet]
         public IActionResult Get() => Json(_options);
 
+        /// <summary>
+        /// Returns generated password as a JSON string.
+        /// </summary>
+        /// <returns>A Json with a password property which contains a random generated password.</returns>
+        [HttpGet]
+        [Route("generated")]
+        public IActionResult GetGeeneratedPassword()
```
Comment (graingert) [inline src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]:
```suggestion
        public IActionResult GetGeneratedPassword()
```

**Comment by graingert** [comment]
could you also add some test cases for this? Eg one for excising the form to show that you can't just POST a new password to the API to replace the generated one.

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs (hunk 1 above)
Comment (graingert) [inline src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: this loop actually reduces the security

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs [hunk 4]
```diff
@@ -43,6 +52,42 @@ public class PasswordController : Controller
...
+            var score = 0;
+            var password = string.Empty;
+            var maxNumber = _words.Length;
+
+            do
+            {
+                var firstIndex = _random.Next(0, maxNumber);
+                var secondIndex = _random.Next(0, maxNumber);
+
+                // if both indexes have the same value, repeat the random
```
Comment (graingert) [inline src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: this check actually reduces security

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs [hunk 5]
```diff
@@ -18,22 +20,29 @@ public class PasswordController : Controller
...
             IOptions<ClientSettings> optionsAccessor,
+            IOptions<Words> wordsAccessor,
             IPasswordChangeProvider passwordChangeProvider)
         {
             _logger = logger;
             _options = optionsAccessor.Value;
+            _words = wordsAccessor.Value.WordsDictionary;
             _passwordChangeProvider = passwordChangeProvider;
+
+            if (_options.ForcePasswordGeneration) _random = new Random();
```
Comment (graingert) [inline src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: use secure random numbers https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rngcryptoserviceprovider?redirectedfrom=MSDN&view=netframework-4.8

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs [hunk 6]
```diff
@@ -43,6 +52,42 @@ public class PasswordController : Controller
         [HttpGet]
         public IActionResult Get() => Json(_options);
 
+        /// <summary>
+        /// Returns generated password as a JSON string.
```
Comment (graingert) [inline src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: This is a really insecure passphrase generator.

If you're going to create human readable/memorable passwords you should use an established and well reviewed algorithm: https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#Generating_the_mnemonic

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs (hunk 1 above)
Comment (graingert) [reply src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: You shouldn't check the score. Instead encode a specified number of bits

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs (hunk 6 above)
Comment (graingert) [reply src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: @Serk352 this looks like a good library https://github.com/elucidsoft/dotnetstandard-bip39

Comment (geoperez) [reply src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: I don't know if we should add BIP39. Maybe keep the current generator and alternative fixed-length random char generator. What do you think?

Comment (graingert) [reply src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: The current generator is insecure and encourages bad passwords, whatever choice is made about BIP39 the current generator must be removed

**Comment by graingert** [comment]
For generating random passwords I'd recommend base85 encoding random bytes retrieved from  RNGCryptoServiceProvider.GetBytes

https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rngcryptoserviceprovider.getbytes?view=netframework-4.8

https://www.nuget.org/packages/SimpleBase/

File: src/Unosquare.PassCore.Web/Controllers/PasswordController.cs (hunk 1 above)
Comment (Serk352) [reply src/Unosquare.PassCore.Web/Controllers/PasswordController.cs]: @graingert can you explain to us more details about the encoding?

**Comment by graingert** [comment]
@geoperez do you know when this will be available in a release?

**Comment by geoperez** [comment]
Today, I just need a couple of minutes.

**Comment by graingert** [comment]
Awesome! Thanks @Serk352 and @geoperez

**Pull request merged**